common_spec_types.go

```diff
@@ -0,0 +1,325 @@
+}
+
+// FlattenSpec is responsible for flattening nested input JSON data into Druid's flat data model.
+type FlattenSpec struct {
+	UseFieldDiscovery bool      `json:"useFieldDiscovery,omitempty"`
+	Fields            FieldList `json:"fields"`
+}
+
+// TimestampSpec is responsible for configuring the primary timestamp.
+type TimestampSpec struct {
+	Column string `json:"column"`
+	Format string `json:"format"`
+}
+
+// FieldList is a list of Fields for ingestion FlattenSpec.
+type FieldList []Field
+
+// Field defines a single filed configuration of the FlattenSpec.
+type Field struct {
+	Type string `json:"type"`
+	Name string `json:"name"`
+	Expr string `json:"expr"`
+}
+
+// Transform defines a single filed transformation of the TransformSpec.
+type Transform struct {
+	Type string `json:"type"`
+	Name string `json:"name"`
+	Expr string `json:"expression"`
+}
+
+// SpatialDimension represents single spatial dimension datum.
+// https://druid.apache.org/docs/latest/querying/geo/#spatial-indexing
+type SpatialDimension struct {
+	DimensionName string   `json:"dimName"`
+	Dimensions    []string `json:"dims,omitempty"`
+}
+
+// TransformSet is a unique set of transforms applied to the input.
+type TransformSet []Transform
+
+// DimensionSet is a unique set of druid datasource dimensions(labels).
+type DimensionSet []any
+
+// Dimension is a typed definition of a datasource dimension.
+type Dimension struct {
+	Type string `json:"type"`
+	Name string `json:"name"`
+}
+
+// SpatialDimensionSet is a unique set of druid datasource spatial dimensions.
+type SpatialDimensionSet []SpatialDimension
+
+// DimensionExclusionsSet represents set of excluded dimensions.
+type DimensionExclusionsSet []string
+
+// DimensionsSpec is responsible for configuring Druid's dimensions. They're a
+// set of columns in Druid's data model that can be used for grouping, filtering
+// or applying aggregations.
+// https://druid.apache.org/docs/latest/ingestion/ingestion-spec#dimensionsspec
+type DimensionsSpec struct {
+	Dimensions           DimensionSet           `json:"dimensions,omitempty"`
+	DimensionExclusions  DimensionExclusionsSet `json:"dimensionExclusions,omitempty"`
+	SpatialDimensions    SpatialDimensionSet    `json:"spatialDimensions,omitempty"`
+	IncludeAllDimensions bool                   `json:"includeAllDimensions,omitempty"`
+	UseSchemaDiscovery   bool                   `json:"useSchemaDiscovery,omitempty"`
+}
+type QueryGranularity struct {
+	Type string `json:"type,omitempty"`
+}
+
+// GranularitySpec allows for configuring operations such as data segment
+// partitioning, truncating timestamps, time chunk segmentation or roll-up.
+// https://druid.apache.org/docs/latest/ingestion/ingestion-spec#granularityspec
+type GranularitySpec struct {
+	Type               string `json:"type"`
+	SegmentGranularity string `json:"segmentGranularity,omitempty"`
+	// TODO: this field is problematic as depending on value druid returns string or object
```

Comment:
Please refer to the comments in https://github.com/h2oai/go-druid/pull/7 also apply here.